### DISCUSSION TITLE
QR code is not working correctly

### DISCUSSION THREAD

**Discussion by Micheal3030** [YOURLS/YOURLS#3819] [Q&A] [answered]
Hello everyone,

We are using YOURLS to create a short URL to use with a static QR code.
The URL can be updated easily, and the QR code will keep the same format.
Therefore, we have a workaround to have a dynamic QR-code under our own management.

The problem we are facing;
After we change the URL, phones keep going to the old one instead of the new one.
When putting the URL into the desktop browser, the correct URL is shown.

After deleting all the website data on the iPhone, the correct URL is still not shown.

Hope anyone got a solution for this. 

Version: 1.9.2

Thank you,

**Comment by Micheal3030** [comment]


**Reply by dgw** [reply]
This is unrelated to the redirection issue, but should be fixed if you care about the access logs.

**Comment by dgw** [comment]
This is unrelated to the QR code. YOURLS by default uses a "permanent" redirect that browsers cache _very_ aggressively.

#3545 describes something similar, as do #1778 and #1832 and many other threads over the years. The solution is a plugin to send temporary redirects instead.

**Reply by Micheal3030** [reply]
Thank you so much for your solution.
Just installed a plugin for the 302 redirect. ([302-instead](https://github.com/EpicPilgrim/302-instead))

But doesn't seem to solve it right away.
Do you know if there is some time needed for the updated URL to be affective?

The strange this is not all devices are facing this problem.
The Phone of my colleague is showing to correct URL while mine is still showing the old one.

Thanks again!!

**Reply by dgw** [reply]
Devices that have never followed this QR code's URL before should be fine going forward, but devices with the permanent redirect cached will still use the cached value for as long as the browser decides to keep it, unless manually cleared.

How to clear the cache on a given device depends on the browser+OS combination. Some are also more stubborn than others, in my experience.

**Reply by Micheal3030** [reply]
Appreciate your fast response and clear answers, great!

So with this 302 plugin the cache of the actual device is not considered?

Sorry if sounds stupid here, just learned about the 302 and 301 redirects.

**Reply by dgw** [reply]
The source of the problem on devices with cached 301s is that they no longer bother contacting the server—YOURLS—because the 301 they have is "permanent". They believe they already have the "correct" response and are following the server's previous instructions to (effectively) "always use this address instead of the one you asked for". The 302 Instead plugin can't override responses already sent before it was installed.

**Reply by Micheal3030** [reply]
Clear enough, thank you.

Tested now with a new link but the same problem occurs, I will try another 302 plugin as this not seem to work properly.

Thanks again for your assistance.

**Reply by Micheal3030** [reply]
The other 302 plugin gives the same issue unfortunately...
Gave it some time as well after changing the URL and testing again, but the same problem.

Do you have any other suggestions maybe?

Thank you in advance,

**Reply by dgw** [reply, marked as answer]
My own testing indicates that browsers will still cache 302s for some period of time without being explicitly told not to. HTTP headers can influence how long it takes; by default, on my YOURLS instance, I see a `max-age` of 600 seconds.

YOURLS includes a function that it uses to send headers disable caching on some pages, mostly in the admin area. I created a simple plugin that calls it also for shorturl redirections: https://github.com/dgw/yourls-shorturl-no-cache

In further testing, I no longer see "(from disk cache)" in Dev Tools' Network pane when accessing recently-used shorturls on my instance, and editing the long URL now takes effect immediately. I've proposed adding the plugin to the awesome list at YOURLS/awesome#183.

**Reply by Micheal3030** [reply]
Installed this plugin you created, and it works great!
Solved the issue, the URL takes effect immediately after change!

Tested on a few different devices and browsers.

Cannot expres how much this is helping us!
Much much appriciated for your effort and keep up the good work :)

For others with this problem, I installed the 302 plugin in combination with the plugin @dgw created.
